allow artifact string macro to take an explicit path to the artifact file

KristofferC opened the pull request:
can be used by JLLs to avoid scouring through the file system for the Artifacts.toml when it already knows where it is

JLLWrappers PR at https://github.com/JuliaPackaging/JLLWrappers.jl/pull/45

Commit:
allow artifact string macro to take an explicit path

can be used by JLLs to avoid scour throgh the file system for the Artifacts.toml when it already knows where it is

staticfloat reviewed (approved):
Can we call the value “artifacts_toml_path” or something like that, to avoid confusion with the actual path of the artifact we want to return?

vtjnash commented on stdlib/Artifacts/test/runtests.jl:
```diff
@@ -91,6 +91,9 @@ end
         HelloWorldC_exe_path = joinpath(HelloWorldC_dir, "bin", "hello_world$(exeext)")
         @test isfile(HelloWorldC_exe_path)
 
+        HelloWorldC_dir_explicit_artifact = eval(Expr(:macrocall, Symbol("@artifact_str"), nothing, "HelloWorldC", nothing, joinpath(@__DIR__, "Artifacts.toml")))
```
`:macrocall` is generally not something that should be constructed by hand. `eval(:(@artifact_str "HelloWorldC" joinpath(@__DIR__, "Artifacts.toml"))` would be a better way to write that

KristofferC replied:
Ok, I just copied https://github.com/JuliaPackaging/JLLWrappers.jl/blob/79226fad1b220ba7a3516a66e76922eb84f293ae/src/wrapper_generators.jl#L17.

Commit:
address review comments

staticfloat committed:
Fix `@artifact_str` invocation

We must provide the `platform` argument here before we provide the
`artifacts_toml_path` argument.

staticfloat committed:
Catch quoted `nothing` values as well

KristofferC commented on stdlib/Artifacts/src/Artifacts.jl:
```diff
@@ -659,10 +659,10 @@ macro artifact_str(name, platform=nothing, artifacts_toml_path=nothing)
         find_artifacts_toml(srcfile)
     else
-        artifacts_toml_path
+        eval(artifacts_toml_path)
```
Did you try this @staticfloat? It seems to error for me:

```
In [3]: using Vulkan_Headers_jll
[ Info: Precompiling Vulkan_Headers_jll [8d446b21-f3ad-5576-a034-752265b9b6f9]
ERROR: LoadError: Evaluation into the closed module `Artifacts` breaks incremental compilation because the side effects will not be permanent. This is likely due to some other module mutating `Artifacts` with `eval` during precompilation - don't do this.
Stacktrace:
 [1] eval
   @ ./boot.jl:370 [inlined]
 [2] eval(x::String)
   @ Artifacts ~/julia/usr/share/julia/stdlib/v1.9/Artifacts/src/Artifacts.jl:3
 [3] var"@artifact_str"(__source__::LineNumberNode, __module__::Module, name::Any, platform::Any, artifacts_toml_path::Any)
   @ Artifacts ~/julia/usr/share/julia/stdlib/v1.9/Artifacts/src/Artifacts.jl:665
```

staticfloat replied:
No, I only tried it with the test set, which doesn't hit this error, apparently.  @vtjnash what is the correct fix here?  I don't really understand how to properly "get" the value of a string at compile-time I guess.

KristofferC replied:
What was it that failed before this change? I tried it with JLLWrappers and it seemed to work ok without the `eval`.
